### PR TITLE
bpo-31904: disable os.popen and impacted test cases on VxWorks

### PR DESCRIPTION
**Pull request by pxinwr**
VxWorks has no user space shell provided so it can't support os.popen(). Disable it on VxWorks and impacted test cases.

https://bugs.python.org/issue41442

### CODE REVIEW

File: Lib/test/test_posix.py [hunk 1]
```diff
@@ -1029,6 +1030,7 @@ def test_getgrouplist(self):
 
 
     @unittest.skipUnless(hasattr(os, 'getegid'), "test needs os.getegid()")
+    @unittest.skipUnless(unix_shell and os.path.exists(unix_shell), 'requires a shell')
```
Comment (ZackerySpytz) [inline Lib/test/test_posix.py]: Please limit lines to 79 characters.

File: Lib/os.py [hunk 2]
```diff
@@ -36,7 +36,9 @@
 __all__ = ["altsep", "curdir", "pardir", "sep", "pathsep", "linesep",
            "defpath", "name", "path", "devnull", "SEEK_SET", "SEEK_CUR",
            "SEEK_END", "fsencode", "fsdecode", "get_exec_path", "fdopen",
-           "popen", "extsep"]
+           "extsep"]
+
+VXWORKS = (sys.platform == 'vxworks')
```
Comment (vstinner) [inline Lib/os.py]: This variable is only at a single place: you can remove it.

File: Lib/os.py [hunk 3]
```diff
@@ -970,50 +972,53 @@ def spawnlpe(mode, file, *args):
...
-        self._stream = stream
-        self._proc = proc
-    def close(self):
-        self._stream.close()
-        returncode = self._proc.wait()
-        if returncode == 0:
-            return None
-        if name == 'nt':
-            return returncode
+if not VXWORKS:
```
Comment (vstinner) [inline Lib/os.py]: I suggest: `if sys.platform != 'vxworks':`.

**Review by vstinner — COMMENTED** [review]
Is this change a temporary workaround until subprocess is fixed on VxWorks, or is this skip supposed to be a temporary workaround?

If it's a temporary workaround, I am not sure if it's a good idea to document the skip.

File: Lib/os.py (hunk 2 above)
Comment (pxinwr) [reply Lib/os.py]: removed.

File: Lib/os.py (hunk 3 above)
Comment (pxinwr) [reply Lib/os.py]: changed accordingly.

**Comment by pxinwr** [comment]
> Is this change a temporary workaround until subprocess is fixed on VxWorks, or is this skip supposed to be a temporary workaround?
> 
> If it's a temporary workaround, I am not sure if it's a good idea to document the skip.

os.popen() runs cmd in shell, i.e. shell=True was set when calling subprocess.Popen(). VxWorks has no user space shell provided so that we can't create a new shell and run commands inside. For the subprocess module, VxWorks can't support features needing shell to run.

File: Lib/test/test_popen.py [hunk 4]
```diff
@@ -7,6 +7,9 @@
 from test import support
 import os, sys
 
+if not hasattr(os, 'popen'):
+    raise unittest.SkipTest("os.popen not defined")
```
Comment (vstinner) [inline Lib/test/test_popen.py]:
```suggestion
    raise unittest.SkipTest("need os.popen()")
```

File: Lib/os.py [hunk 5]
```diff
@@ -970,50 +970,53 @@ def spawnlpe(mode, file, *args):
...
-        self._stream = stream
-        self._proc = proc
-    def close(self):
-        self._stream.close()
-        returncode = self._proc.wait()
-        if returncode == 0:
-            return None
-        if name == 'nt':
-            return returncode
+if sys.platform != 'vxworks':
```
Comment (vstinner) [inline Lib/os.py:974]: Please add a comment explaining why it's not available on VxWorks. I understood that the issue is that VxWorks doesn't support Popen(shell=True).

Comment (pxinwr) [reply Lib/os.py:974]: Added.

**Comment by vstinner** [comment]
Merged, thanks.

**Pull request merged**